src/libraries/System.Private.CoreLib/src/System/Text/Latin1Utility.cs

```diff
@@ -0,0 +1,1101 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Runtime.Intrinsics;
+using System.Runtime.Intrinsics.X86;
+using Internal.Runtime.CompilerServices;
+
+#pragma warning disable SA1121 // explicitly using type aliases instead of built-in types
+#if TARGET_64BIT
+using nint = System.Int64;
+using nuint = System.UInt64;
+#else // TARGET_64BIT
+using nint = System.Int32;
+using nuint = System.UInt32;
+#endif // TARGET_64BIT
+
+namespace System.Text
+{
+    internal static partial class Latin1Utility
+    {
+#if DEBUG
+        static Latin1Utility()
+        {
+            Debug.Assert(sizeof(nint) == IntPtr.Size && nint.MinValue < 0, "nint is defined incorrectly.");
+            Debug.Assert(sizeof(nuint) == IntPtr.Size && nuint.MinValue == 0, "nuint is defined incorrectly.");
+        }
+#endif // DEBUG
+
+        /// <summary>
+        /// Returns the index in <paramref name="pBuffer"/> where the first non-Latin1 char is found.
+        /// Returns <paramref name="bufferLength"/> if the buffer is empty or all-Latin1.
+        /// </summary>
+        /// <returns>A Latin-1 char is defined as 0x0000 - 0x00FF, inclusive.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe nuint GetIndexOfFirstNonLatin1Char(char* pBuffer, nuint bufferLength /* in chars */)
+        {
+            // If SSE2 is supported, use those specific intrinsics instead of the generic vectorized
+            // code below. This has two benefits: (a) we can take advantage of specific instructions like
+            // pmovmskb which we know are optimized, and (b) we can avoid downclocking the processor while
+            // this method is running.
+
+            return (Sse2.IsSupported)
+                ? GetIndexOfFirstNonLatin1Char_Sse2(pBuffer, bufferLength)
+                : GetIndexOfFirstNonLatin1Char_Default(pBuffer, bufferLength);
+        }
+
+        private static unsafe nuint GetIndexOfFirstNonLatin1Char_Default(char* pBuffer, nuint bufferLength /* in chars */)
+        {
+            // Squirrel away the original buffer reference.This method works by determining the exact
+            // char reference where non-Latin1 data begins, so we need this base value to perform the
+            // final subtraction at the end of the method to get the index into the original buffer.
+
+            char* pOriginalBuffer = pBuffer;
+
+            Debug.Assert(bufferLength <= nuint.MaxValue / sizeof(char));
+
+            // Before we drain off char-by-char, try a generic vectorized loop.
+            // Only run the loop if we have at least two vectors we can pull out.
+
+            if (Vector.IsHardwareAccelerated && bufferLength >= 2 * (uint)Vector<ushort>.Count)
```

Comment:
This uses `Vector<T>` and I noticed you called out wanting to use instructions like `pmovmskb`.

Do you still get similar wins if you utilize the `Vector<T>` to/from `Vector128/256<T>` conversion APIs for just when those instructions are needed?

Reply:
I didn't try this, but I'm not quite sure how to do this cleanly. It's not just _pmovmskb_. I'm also using _ptest_ (if allowed) and _paddusw_ (if _ptest_ is unavailable). I also have different alignment requirements between the two methods: the SSE2-specific code path uses only 128-bit instructions because I didn't see any perf gains (and in fact saw a loss for small inputs) when switching to 256-bit instructions. This also means that the SSE2-specific code paths only require 128-bit alignment, whereas the generalized code path has no particular alignment requirements.